File: legacy/source/extensions/isaacsim.core.cloner/python/impl/grid_cloner.py
# SPDX-FileCopyrightText: Copyright (c) 2022-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Grid cloner module."""

from __future__ import annotations

__all__ = ["GridCloner"]


import numpy as np
from isaacsim.core.cloner import Cloner
from pxr import Gf, Usd, UsdGeom


class GridCloner(Cloner):
    """A specialized Cloner class that automatically generates clones in a grid pattern.

    This class extends :class:`Cloner` to provide automatic grid-based positioning
    of clones, simplifying the creation of environments arranged in a regular grid.

    Args:
        spacing: Spacing between clones in the grid.
        num_per_row: Number of clones to place in a row. Defaults to sqrt(num_clones)
            if set to -1.
        stage: USD stage where source prim and clones are added to.
            Defaults to the current stage from the USD context.

    Example:

    .. code-block:: python

        >>> from isaacsim.core.cloner import GridCloner
        >>>
        >>> cloner = GridCloner(spacing=2.0)
        >>> cloner.define_base_env("/World/envs")
        >>> prim_paths = cloner.generate_paths("/World/envs/env", 9)
        >>> positions = cloner.clone(
        ...     source_prim_path="/World/envs/env_0",
        ...     prim_paths=prim_paths,
        ... )

    .. code-block:: python

        >>> from isaacsim.core.cloner import GridCloner
        >>>
        >>> # Create a grid cloner with 2.0 spacing
        >>> cloner = GridCloner(spacing=2.0)
        >>>
        >>> # Create a grid cloner with 3 clones per row
        >>> cloner = GridCloner(spacing=1.5, num_per_row=3)
    """

    def __init__(self, spacing: float, num_per_row: int = -1, stage: Usd.Stage = None) -> None:
        self._spacing = spacing
        self._num_per_row = num_per_row

        self._positions = None
        self._orientations = None
        self._cached_num_clones = None

        Cloner.__init__(self, stage)

    def get_clone_transforms(
        self,
        num_clones: int,
        position_offsets: np.ndarray = None,
        orientation_offsets: np.ndarray = None,
    ) -> tuple[list, list]:
        """Compute the positions and orientations of clones in a grid.

        Args:
            num_clones: Number of clones.
            position_offsets: Positions to be applied as local translations on top of
                computed clone position. Defaults to None, no offset will be applied.
            orientation_offsets: Orientations to be applied as local rotations for each
                clone as quaternions (w, x, y, z). Defaults to None, no offset will be applied.

        Returns:
            A tuple containing:
                - positions: Computed positions of all clones.
                - orientations: Computed orientations of all clones as quaternions (w, x, y, z).

        Raises:
            ValueError: If the dimension of position_offsets does not match num_clones.
            ValueError: If the dimension of orientation_offsets does not match num_clones.

        Example:

        .. code-block:: python

            >>> import numpy as np
            >>> from isaacsim.core.cloner import GridCloner
            >>>
            >>> cloner = GridCloner(spacing=2.0)
            >>> positions, orientations = cloner.get_clone_transforms(num_clones=4)
            >>> len(positions)
            4
        """
        # check if inputs are valid
        if position_offsets is not None:
            if len(position_offsets) != num_clones:
                raise ValueError("Dimension mismatch between position_offsets and prim_paths!")
            # convert to numpy array
            # - convert from torch (without explicit importing it)
            try:
                position_offsets = position_offsets.detach().cpu().numpy()
            except Exception:
                pass
            # - convert from other types
            if not isinstance(position_offsets, np.ndarray):
                position_offsets = np.asarray(position_offsets)
        if orientation_offsets is not None:
            if len(orientation_offsets) != num_clones:
                raise ValueError("Dimension mismatch between orientation_offsets and prim_paths!")
            # convert to numpy array
            # - convert from torch (without explicit importing it)
            try:
                orientation_offsets = orientation_offsets.detach().cpu().numpy()
            except Exception:
                pass
            # - convert from other types
            if not isinstance(orientation_offsets, np.ndarray):
                orientation_offsets = np.asarray(orientation_offsets)

        if num_clones < 0:
            raise ValueError("num_clones must be non-negative")

        use_cache = position_offsets is None and orientation_offsets is None

        if (
            use_cache
            and self._positions is not None
            and self._orientations is not None
            and self._cached_num_clones == num_clones
        ):
            return self._positions, self._orientations

        if num_clones == 0:
            positions = []
            orientations = []
            if use_cache:
                self._positions = positions
                self._orientations = orientations
                self._cached_num_clones = num_clones
            return positions, orientations

        num_per_row = int(np.sqrt(num_clones)) if self._num_per_row == -1 else self._num_per_row
        if num_per_row <= 0:
            raise ValueError("num_per_row must be positive")

        num_rows = np.ceil(num_clones / num_per_row)
        num_cols = np.ceil(num_clones / num_rows)

        row_offset = 0.5 * self._spacing * (num_rows - 1)
        col_offset = 0.5 * self._spacing * (num_cols - 1)

        positions = []
        orientations = []

        for i in range(num_clones):
            # compute transform
            row = i // num_cols
            col = i % num_cols
            x = row_offset - row * self._spacing
            y = col * self._spacing - col_offset

            up_axis = UsdGeom.GetStageUpAxis(self._stage)
            position = [x, y, 0] if up_axis == UsdGeom.Tokens.z else [x, 0, y]
            orientation = Gf.Quatd.GetIdentity()

            if position_offsets is not None:
                translation = position_offsets[i] + position
            else:
                translation = position

            if orientation_offsets is not None:
                orientation = (
                    Gf.Quatd(orientation_offsets[i][0].item(), Gf.Vec3d(orientation_offsets[i][1:].tolist()))
                    * orientation
                )

            orientation = [
                orientation.GetReal(),
                orientation.GetImaginary()[0],
                orientation.GetImaginary()[1],
                orientation.GetImaginary()[2],
            ]

            positions.append(translation)
            orientations.append(orientation)

        if use_cache:
            self._positions = positions
            self._orientations = orientations
            self._cached_num_clones = num_clones

        return positions, orientations

    def clone(
        self,
        source_prim_path: str,
        prim_paths: list[str],
        position_offsets: np.ndarray = None,
        orientation_offsets: np.ndarray = None,
        replicate_physics: bool = False,
        base_env_path: str = None,
        root_path: str = None,
        copy_from_source: bool = False,
        enable_env_ids: bool = False,
        clone_in_fabric: bool = False,
    ) -> list:
        """Create clones in a grid pattern with automatically computed positions.

        Args:
            source_prim_path: Path of the source object.
            prim_paths: List of destination paths.
            position_offsets: Positions to be applied as local translations on top of
                computed clone position. Defaults to None, no offset will be applied.
            orientation_offsets: Orientations to be applied as local rotations for each
                clone as quaternions (w, x, y, z). Defaults to None, no offset will be applied.
            replicate_physics: Uses omni.physics replication. This will replicate physics
                properties directly for paths beginning with root_path and skip physics
                parsing for anything under the base_env_path.
            base_env_path: Path to namespace for all environments. Required if
                replicate_physics=True and define_base_env() was not called.
            root_path: Prefix path for each environment. Required if replicate_physics=True
                and generate_paths() was not called.
            copy_from_source: Setting this to False will inherit all clones from the source
                prim; any changes made to the source prim will be reflected in the clones.
                Setting this to True will make copies of the source prim when creating new
                clones; changes to the source prim will not be reflected in clones.
                Defaults to False. Note that setting this to True will take longer to execute.
            enable_env_ids: Setting this enables co-location of clones in physics with
                automatic filtering of collisions between clones.
            clone_in_fabric: Whether to perform cloning operations in Fabric for improved
                performance.

        Returns:
            Computed positions of all clones.

        Example:

        .. code-block:: python

            >>> from isaacsim.core.cloner import GridCloner
            >>>
            >>> cloner = GridCloner(spacing=2.0)
            >>> cloner.define_base_env("/World/envs")
            >>> prim_paths = cloner.generate_paths("/World/envs/env", 9)
            >>> positions = cloner.clone(
            ...     source_prim_path="/World/envs/env_0",
            ...     prim_paths=prim_paths,
            ... )
            >>> len(positions)
            9
        """
        num_clones = len(prim_paths)

        positions, orientations = self.get_clone_transforms(num_clones, position_offsets, orientation_offsets)

        super().clone(
            source_prim_path=source_prim_path,
            prim_paths=prim_paths,
            positions=positions,
            orientations=orientations,
            replicate_physics=replicate_physics,
            base_env_path=base_env_path,
            root_path=root_path,
            copy_from_source=copy_from_source,
            enable_env_ids=enable_env_ids,
            clone_in_fabric=clone_in_fabric,
        )

        return positions

File: crates/isaacsim-core-cloner/src/grid_cloner.rs
// SPDX-FileCopyrightText: Copyright (c) 2022-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Grid cloner module.
//!
//! Port of `python/impl/grid_cloner.py` from the legacy `isaacsim.core.cloner`
//! extension.

use isaacsim_scene::{SceneStage, UpAxis};

use crate::cloner::{CloneOptions, Cloner};

/// A specialized [`Cloner`] that automatically generates clones in a grid
/// pattern.
///
/// `num_per_row` defaults to `sqrt(num_clones)` when `None` (legacy `-1`).
///
/// Note: deliberately not `Clone` — the legacy-named [`Self::clone`] prim
/// duplication method would otherwise be shadowed by `Clone::clone`.
#[derive(Debug)]
pub struct GridCloner {
    cloner: Cloner,
    spacing: f64,
    num_per_row: Option<usize>,
    cached_positions: Option<Vec<[f64; 3]>>,
    cached_orientations: Option<Vec<[f64; 4]>>,
    cached_num_clones: Option<usize>,
}

impl GridCloner {
    /// Create a grid cloner with the given spacing between clones.
    pub fn new(spacing: f64, num_per_row: Option<usize>) -> Self {
        Self {
            cloner: Cloner::new(),
            spacing,
            num_per_row,
            cached_positions: None,
            cached_orientations: None,
            cached_num_clones: None,
        }
    }

    /// Access the underlying [`Cloner`] (e.g. for `define_base_env`).
    pub fn cloner_mut(&mut self) -> &mut Cloner {
        &mut self.cloner
    }

    /// Create a `Scope` prim at `base_env_path` (see [`Cloner::define_base_env`]).
    pub fn define_base_env<S: SceneStage + ?Sized>(
        &mut self,
        stage: &mut S,
        base_env_path: &str,
    ) -> Result<(), String> {
        self.cloner.define_base_env(stage, base_env_path)
    }

    /// Generate `num_paths` paths in the format `{root_path}_{i}`
    /// (see [`Cloner::generate_paths`]).
    pub fn generate_paths(&mut self, root_path: &str, num_paths: usize) -> Vec<String> {
        self.cloner.generate_paths(root_path, num_paths)
    }

    /// Compute the positions and orientations of clones in a grid.
    ///
    /// `position_offsets` / `orientation_offsets` (quaternions `[w, x, y, z]`)
    /// are applied per clone on top of the computed grid transform. Plain
    /// grid transforms (no offsets) are cached per clone count, as in the
    /// legacy implementation.
    ///
    /// Errors if an offset array length does not match `num_clones` or
    /// `num_per_row` is zero.
    #[allow(clippy::type_complexity)]
    pub fn get_clone_transforms<S: SceneStage + ?Sized>(
        &mut self,
        stage: &S,
        num_clones: usize,
        position_offsets: Option<&[[f64; 3]]>,
        orientation_offsets: Option<&[[f64; 4]]>,
    ) -> Result<(Vec<[f64; 3]>, Vec<[f64; 4]>), String> {
        if let Some(offsets) = position_offsets {
            if offsets.len() != num_clones {
                return Err("Dimension mismatch between position_offsets and prim_paths!".to_string());
            }
        }
        if let Some(offsets) = orientation_offsets {
            if offsets.len() != num_clones {
                return Err("Dimension mismatch between orientation_offsets and prim_paths!".to_string());
            }
        }

        let use_cache = position_offsets.is_none() && orientation_offsets.is_none();
        if use_cache && self.cached_num_clones == Some(num_clones) {
            if let (Some(positions), Some(orientations)) =
                (&self.cached_positions, &self.cached_orientations)
            {
                return Ok((positions.clone(), orientations.clone()));
            }
        }

        if num_clones == 0 {
            if use_cache {
                self.cached_positions = Some(Vec::new());
                self.cached_orientations = Some(Vec::new());
                self.cached_num_clones = Some(0);
            }
            return Ok((Vec::new(), Vec::new()));
        }

        let num_per_row = match self.num_per_row {
            Some(n) => n,
            None => (num_clones as f64).sqrt() as usize,
        };
        if num_per_row == 0 {
            return Err("num_per_row must be positive".to_string());
        }
        let num_rows = num_clones.div_ceil(num_per_row);
        let num_cols = num_clones.div_ceil(num_rows);

        let row_offset = 0.5 * self.spacing * (num_rows - 1) as f64;
        let col_offset = 0.5 * self.spacing * (num_cols - 1) as f64;

        let mut positions = Vec::with_capacity(num_clones);
        let mut orientations = Vec::with_capacity(num_clones);
        for i in 0..num_clones {
            let row = i / num_cols;
            let col = i % num_cols;
            let x = row_offset - row as f64 * self.spacing;
            let y = col as f64 * self.spacing - col_offset;

            let position = match stage.up_axis() {
                UpAxis::Z => [x, y, 0.0],
                UpAxis::Y => [x, 0.0, y],
            };
            let translation = match position_offsets {
                Some(offsets) => [
                    offsets[i][0] + position[0],
                    offsets[i][1] + position[1],
                    offsets[i][2] + position[2],
                ],
                None => position,
            };
            // offset * identity = offset
            let orientation = match orientation_offsets {
                Some(offsets) => offsets[i],
                None => [1.0, 0.0, 0.0, 0.0],
            };

            positions.push(translation);
            orientations.push(orientation);
        }

        if use_cache {
            self.cached_positions = Some(positions.clone());
            self.cached_orientations = Some(orientations.clone());
            self.cached_num_clones = Some(num_clones);
        }
        Ok((positions, orientations))
    }

    /// Create clones in a grid pattern with automatically computed positions.
    ///
    /// Returns the computed positions of all clones.
    pub fn clone<S: SceneStage + ?Sized>(
        &mut self,
        stage: &mut S,
        source_prim_path: &str,
        prim_paths: &[String],
        position_offsets: Option<&[[f64; 3]]>,
        orientation_offsets: Option<&[[f64; 4]]>,
        copy_from_source: bool,
    ) -> Result<Vec<[f64; 3]>, String> {
        let (positions, orientations) =
            self.get_clone_transforms(stage, prim_paths.len(), position_offsets, orientation_offsets)?;
        self.cloner.clone(
            stage,
            source_prim_path,
            prim_paths,
            &CloneOptions {
                positions: Some(positions.clone()),
                orientations: Some(orientations),
                copy_from_source,
            },
        )?;
        Ok(positions)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use isaacsim_scene::{Stage, Value};

    /// Port of legacy `tests/test_cloner.py::test_grid_cloner`.
    #[test]
    fn test_grid_cloner() {
        let mut stage = Stage::new();
        stage.define_prim("/World/Cube_0", "Cube").unwrap();

        let mut cloner = GridCloner::new(3.0, None);
        let target_paths = cloner.generate_paths("/World/Cube", 4);
        cloner
            .clone(&mut stage, "/World/Cube_0", &target_paths, None, None, false)
            .unwrap();

        let target_translations = [
            [1.5, -1.5, 0.0],
            [1.5, 1.5, 0.0],
            [-1.5, -1.5, 0.0],
            [-1.5, 1.5, 0.0],
        ];
        for (i, translation) in target_translations.iter().enumerate() {
            let path = format!("/World/Cube_{i}");
            assert!(stage.prim_exists(&path));
            assert_eq!(stage.composed_type_name(&path).unwrap(), "Cube");
            assert_eq!(
                stage.resolve_attribute(&path, "xformOp:translate"),
                Some(Value::Vec3d(*translation))
            );
        }
    }

    /// Port of legacy `test_grid_cloner_recomputes_transforms_when_count_changes`.
    #[test]
    fn test_grid_cloner_recomputes_transforms_when_count_changes() {
        let mut stage = Stage::new();
        stage.set_up_axis(UpAxis::Z);

        let mut cloner = GridCloner::new(2.0, None);
        let (positions_4, _) = cloner.get_clone_transforms(&stage, 4, None, None).unwrap();
        let (positions_9, _) = cloner.get_clone_transforms(&stage, 9, None, None).unwrap();

        assert_eq!(positions_4.len(), 4);
        assert_eq!(positions_9.len(), 9);
        assert_eq!(positions_9[0], [2.0, -2.0, 0.0]);
        assert_eq!(positions_9[8], [-2.0, 2.0, 0.0]);
    }

    /// Port of legacy `test_grid_cloner_empty_clone_count_returns_empty_transforms`.
    #[test]
    fn test_grid_cloner_empty_clone_count_returns_empty_transforms() {
        let stage = Stage::new();
        let mut cloner = GridCloner::new(2.0, None);
        let (positions, orientations) = cloner.get_clone_transforms(&stage, 0, None, None).unwrap();
        assert!(positions.is_empty());
        assert!(orientations.is_empty());
    }

    /// Port of legacy `test_grid_cloner_offsets`.
    #[test]
    fn test_grid_cloner_offsets() {
        let mut stage = Stage::new();
        stage.define_prim("/World/envs/env_0", "Xform").unwrap();
        stage.define_prim("/World/envs/env_0/Ant", "Xform").unwrap();

        let mut cloner = GridCloner::new(3.0, None);
        cloner.define_base_env(&mut stage, "/World/envs").unwrap();
        let target_paths = cloner.generate_paths("/World/envs/env", 100);

        let position_offsets = vec![[0.0, 0.0, 1.0]; 100];
        let orientation_offsets = vec![[0.0, 0.0, 0.0, 1.0]; 100];
        let target_translations = cloner
            .clone(
                &mut stage,
                "/World/envs/env_0",
                &target_paths,
                Some(&position_offsets),
                Some(&orientation_offsets),
                false,
            )
            .unwrap();

        for (i, translation) in target_translations.iter().enumerate() {
            let path = format!("/World/envs/env_{i}");
            assert!(stage.prim_exists(&path));
            assert!(stage.prim_exists(&format!("{path}/Ant")));
            assert_eq!(
                stage.resolve_attribute(&path, "xformOp:translate"),
                Some(Value::Vec3d(*translation))
            );
            assert_eq!(translation[2], 1.0);
            assert_eq!(
                stage.resolve_attribute(&path, "xformOp:orient"),
                Some(Value::Quatd([0.0, 0.0, 0.0, 1.0]))
            );
        }
    }

    /// Grid layout in a Y-up stage places clones in the XZ plane.
    #[test]
    fn test_grid_cloner_y_up() {
        let mut stage = Stage::new();
        stage.set_up_axis(UpAxis::Y);
        let mut cloner = GridCloner::new(3.0, None);
        let (positions, _) = cloner.get_clone_transforms(&stage, 4, None, None).unwrap();
        assert_eq!(positions[0], [1.5, 0.0, -1.5]);
        assert!(positions.iter().all(|p| p[1] == 0.0));
    }

    #[test]
    fn test_offset_dimension_mismatch_fails() {
        let stage = Stage::new();
        let mut cloner = GridCloner::new(2.0, None);
        assert!(cloner
            .get_clone_transforms(&stage, 4, Some(&[[0.0; 3]; 2]), None)
            .is_err());
        assert!(cloner
            .get_clone_transforms(&stage, 4, None, Some(&[[1.0, 0.0, 0.0, 0.0]; 3]))
            .is_err());
    }

    #[test]
    fn test_explicit_num_per_row() {
        let stage = Stage::new();
        // 4 clones, 1 per row: a single column spread along x
        let mut cloner = GridCloner::new(2.0, Some(1));
        let (positions, _) = cloner.get_clone_transforms(&stage, 4, None, None).unwrap();
        assert_eq!(positions[0], [3.0, 0.0, 0.0]);
        assert_eq!(positions[3], [-3.0, 0.0, 0.0]);
    }
}
